pkg/tracker/uniswapTWAP.go

```diff
@@ -0,0 +1,232 @@
+// Copyright (c) The Tellor Authors.
+// Licensed under the MIT License.
+
+package tracker
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math"
+	"math/big"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+
+	tellorCommon "github.com/tellor-io/telliot/pkg/common"
+	"github.com/tellor-io/telliot/pkg/db"
+	"github.com/tellor-io/telliot/pkg/rpc"
+	uniswapcontract "github.com/tellor-io/telliot/pkg/tracker/uniswap"
+)
+
+// UniswapPriceCumulativeLast will be used to save/load data into/from the DB, so that
+// we can minimize onchain contract calls.
+type UniswapPriceCumulativeLast struct {
+	Price0CumulativeLast string
+	Decimals             uint8
+	Decimals0            uint8
+	Decimals1            uint8
+	TimestampLast        uint32
+	CalculatedPrice0Last float64
+}
+
+// UniswapGetter implements DataSource interface.
+type UniswapGetter struct {
+	pair    string
+	address string
+	client  rpc.ETHClient
+}
+
+func (u *UniswapGetter) String() string {
+	return "UniswapGetter"
+}
+
+func (u *UniswapGetter) dBCumulativePriceKey() string {
+	return fmt.Sprintf("uniswap-%s-cumulative", u.pair)
+}
+
+// NewUniswapGetter creates new UniswapGetter for provided pair and pair address.
+func NewUniswapGetter(pair string, address string) *UniswapGetter {
+	pairAddress := strings.Split(address, ":")[1]
+	return &UniswapGetter{
+		pair:    pair,
+		address: pairAddress,
+	}
+}
+
+// Get calculates price for the provided pair.
+func (u *UniswapGetter) Get(ctx context.Context) ([]byte, error) {
+	//cast client using type assertion since context holds generic interface{}
+	u.client = ctx.Value(tellorCommon.ClientContextKey).(rpc.ETHClient)
+	var err error
+	DB := ctx.Value(tellorCommon.DBContextKey).(db.DB)
+
+	// getting price cumulative last onchain.
+	priceCumulativeLast, err := u.getCumulativePriceLast()
+	if err != nil {
+		return nil, errors.Wrap(err, "getting price cumulative last")
+	}
+	// Try to get previous price0Cumulative, timestamp from DB so that we minimize contract calls.
+	var priceCumulative *UniswapPriceCumulativeLast
+	priceCumulative, err = u.pullPriceFromDB(DB)
+	if err != nil {
+		return nil, errors.Wrap(err, "getting price cumulative from DB")
+	}
+	// Return the price data based on conditions.
+	var priceData []float64
+	if priceCumulative == nil {
+		priceData = []float64{0}
```

Comment:
This doesn't look right - when the db doesn't return any price it should fall back to get it from uniswap right?

Reply:
This is needed for calculating a TWAP price according to Uniswap docs. 
In the time when there is no previous price, we can either return zero and calculate the price on the next call or look back into previous blocks and pick another block for our calculations! 

Which one do you think is best?